File: day2/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LINE_SIZE 1000

FILE *fopen(const char *FileName, const char *OpenMode);

int is_safe(int *line, int size) {
  int direction = 0;
  for (int j = 0; j < size - 1; j++) {
    int diff = line[j + 1] - line[j];

    if (diff != 1 && diff != 2 && diff != 3 && diff != -1 && diff != -2 &&
        diff != -3) {
      return 0;
    }

    if (direction == 0) {
      direction = (diff > 0) ? 1 : (diff < 0) ? -1 : 0;
    } else if ((direction == 1 && diff < 0) || (direction == -1 && diff > 0)) {
      return 0;
    }
  }
  return 1;
}

int safe_line(int **table, int *size, int num_rows) {
  int i = 0, j = 0, safe_line = 0;
  for (i = 0; i < num_rows; i++) {
    if (is_safe(table[i], size[i])) {
      safe_line++;
    } else {
      int safe = 0;
      for (j = 0; j < size[i]; j++) {
        int *modified_line = malloc((size[i]) * sizeof(int));
        int a = 0;

        for (int b = 0; b < size[i]; b++) {
          if (b != j) {
            modified_line[a++] = table[i][b];
          }
        }
        if (is_safe(modified_line, size[i] - 1)) {
          safe = 1;
          free(modified_line);
          break;
        }

        free(modified_line);
      }
      if (safe) {
        safe_line++;
      }
    }
  }
  return safe_line;
}

int main(void) {

  FILE *file = NULL;
  int i = 0;

  char line[MAX_LINE_SIZE];
  int **table = NULL;
  int *size = NULL;
  int num_rows = 0;

  file = fopen("../input.txt", "r");
  if (file != NULL) {
    while (fgets(line, sizeof(line), file)) {
      int count = 0;
      char *temp = strdup(line);
      char *token = strtok(temp, " ");
      while (token != NULL) {
        count++;
        token = strtok(NULL, " ");
      }
      free(temp);

      size = (int *)realloc(size, (num_rows + 1) * sizeof(int));
      table = (int **)realloc(table, (num_rows + 1) * sizeof(int *));

      if (size == NULL || table == NULL) {
        printf("Error memory allocation");
        return 1;
      }

      size[num_rows] = count;
      table[num_rows] = (int *)malloc(count * sizeof(int));
      if (table[num_rows] == NULL) {
        printf("Error memory allocation");
        return 1;
      }
      int i = 0;
      token = strtok(line, " ");
      while (token != NULL) {
        table[num_rows][i++] = atoi(token);
        token = strtok(NULL, " ");
      }
      num_rows++;
    }
    fclose(file);
  } else {
    printf("error file");
  }

  int safe_line_total = safe_line(table, size, num_rows);
  printf("%d\n", safe_line_total);

  for (int i = 0; i < num_rows; i++) {
    free(table[i]);
  }
  free(table);
  free(size);

  return 0;
}

File: day2/rust/src/main.rs
use std::fs::File;
use std::io::{self, BufRead, BufReader};

fn is_safe(line: &[i32]) -> bool {
    let size = line.len();
    if size < 2 {
        return true;
    }
    let mut direction = 0;
    for element in 0..(size - 1) {
        let diff = line[element + 1] - line[element];
        if diff.abs() > 3 || diff.abs() < 1 {
            return false;
        }
        if direction == 0 {
            direction = match diff.cmp(&0) {
                std::cmp::Ordering::Greater => 1,
                std::cmp::Ordering::Less => -1,
                std::cmp::Ordering::Equal => 0,
            };
        } else if (direction == 1 && diff < 0) || (direction == -1 && diff > 0) {
            return false;
        }
    }
    true
}

fn safe_line(table: &[Vec<i32>], sizes: &[usize], num_rows: usize) -> i32 {
    let mut safe_line_count = 0;

    for i in 0..num_rows {
        let line = &table[i];
        let size = sizes[i];
        if is_safe(line) {
            safe_line_count += 1;
        } else {
            let mut safe = false;
            for j in 0..size {
                let mut modified_line = line.clone();
                modified_line.remove(j);
                if is_safe(&modified_line) {
                    safe = true;
                    break;
                }
            }
            if safe {
                safe_line_count += 1;
            }
        }
    }

    safe_line_count
}

fn main() -> io::Result<()> {
    let file = File::open("day2/input.txt")?;
    let reader = BufReader::new(file);

    let mut table: Vec<Vec<i32>> = Vec::new();
    let mut sizes: Vec<usize> = Vec::new();

    for line in reader.lines() {
        let line = line?;
        let mut row: Vec<i32> = Vec::new();

        for token in line.split_whitespace() {
            if let Ok(num) = token.parse::<i32>() {
                row.push(num);
            } else {
                eprintln!("Error token: {}", token);
            }
        }
        sizes.push(row.len());
        table.push(row);
    }
    let num_rows = table.len();
    let safe_line_total = safe_line(&table, &sizes, num_rows);
    println!("result {}", safe_line_total);

    Ok(())
}
